extensions/amp-form/0.1/amp-form.js

```diff
@@ -721,42 +724,78 @@ function isDisabled_(element) {
 
 
 /**
- * Installs submission handler on all forms in the document.
- * @param {!Window} win
+ * Bootstraps the amp-form elements
  */
-function installSubmissionHandlers(win) {
-  onDocumentReady(win.document, doc => {
-    toArray(doc.forms).forEach((form, index) => {
+export class AmpFormService {
+  /**
+   * @param  {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @const {!Document|!ShadowRoot} */
+    this.rootNode_ = ampdoc.getRootNode();
+
+    /** @const {boolean} */
+    this.isShadow_ = isShadowRoot(this.rootNode_);
+
+    this.initialize_();
+  }
+
+  /**
+   * Install the amp-form CSS and handlers on the ampdoc.
+   */
+  initialize_() {
+    if (this.isShadow_) {
+      const root = /** @type {!ShadowRoot} */ (this.rootNode_);
+      installStylesForShadowRoot(root, CSS);
+      this.install_();
+    } else {
+      const root = /** @type {!Document} */ (this.rootNode_);
+      installStyles(root, CSS, () => this.install_());
+    }
+  }
+
+  /**
+   * Install the appropriate handlers depending on the type of ampdoc.
+   */
+  install_() {
+    if (this.isShadow_) {
```

Comment:
In both cases, we need to do this:
```
ampdoc.whenReady().then(() => {
  installHandlers(ampdoc.getRootNode().querySelectorAll('form'));
});
```

Overall, the desire is to minimize `if`s for shadow-vs-single-doc cases. Really, I believe, CSS creation should be the only needed distinction. So much so, I'd recommend to remove `this.isShadow_` private var to remove incentive to use it.